feat: add terraform workspace segment

matsest opened the pull request:
### Prerequisites

- [x] I have read and understand the `CONTRIBUTING` guide
- [x] The commit message follows the [conventional commits][cc] guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

### Description

This adds support for showing the [current terraform workspace name](https://www.terraform.io/docs/commands/workspace/show.html).

Inspired by a bash prompt I've used to show the current workspace:

```
function terraform_prompt()
  {
     if [ -d .terraform ]; then
       workspace="$(command terraform workspace show 2>/dev/null)"
       echo " (${workspace})"
     fi
  }
```

[cc]: https://www.conventionalcommits.org/en/v1.0.0/#summary

matsest committed:
feat: add terraform workspace segment

matsest commented:
The existence of the .terraform dir and test for it is the WIP part I've identified and not too sure about (not very familiar with golang)

Edit: I think this is working now, after testing locally and running the gh actions.  But very open to improvement suggestions :)

JanDeDobbeleer commented on docs/docs/segment-terraform.md:
```diff
@@ -0,0 +1,24 @@
+{
+  "type": "terraform",
+  "style": "powerline",
+  "powerline_symbol": "",
```
Can you change this to `"powerline_symbol": "\uE0B0",`, I changed all docs to use UTF32 text representations rather than icons which can't always be rendered

JanDeDobbeleer commented on segment_terraform.go:
```diff
@@ -0,0 +1,24 @@
+}
+
+func (tf *terraform) enabled() bool {
+	if !(tf.env.hasCommand("terraform") && tf.env.hasFolder(".terraform")) {
```
Can we adjust this to `!tf.env.hasCommand("terraform") && !tf.env.hasFolder(".terraform")`, when I read this I didn't see the initial `!` and I'm sure I won't be the only one :-)

matsest replied:
Done :)

matsest replied:
Actually, it seems like if !(A && B) and  if !A && !B does not evaluate in the same manner.

With the first implementation (how it was) it did not show the prompt if there was not a .terraform directory, but in the suggested implementation it actually shows the prompt despite not having a .terraform directory. I've tested it with the current build.

Is this a go-ism? @JanDeDobbeleer  Agree that for readability the suggested implementation is better.

Consider this example where the third statement is equivalent to what happens now:
```
func main() {
        if !false && !false {
                fmt.Println("!false && !false works")
        }

        if !(false && false) {
                fmt.Println("!(false && false) works")
        }

        if !true && !false {
                fmt.Println("!true && !false works") <------ this output will not be printed.
        }

        if !(true && false) {
                fmt.Println("!(true && false) works")
        }
}
``` 

This is however fixable with an || statement.. (if !A || !B). But I think that is a bit less clear. Let me know what you think.

matsest replied:
A suggestion for how to include this test into segment_terraform_test.go (actually evaluate both cases together) would be of great help! :) I started - see b413a7c

JanDeDobbeleer replied:
It should be `!tf.env.hasCommand("terraform") || !tf.env.hasFolder(".terraform")`, when either of one condition isn't met, it's disabled.

JanDeDobbeleer commented on segment_terraform.go:
```diff
@@ -0,0 +1,24 @@
+	if !(tf.env.hasCommand("terraform") && tf.env.hasFolder(".terraform")) {
+		return false
+	}
+	tf.workspaceName = tf.env.runCommand("terraform", "workspace", "show")
```
Can this fail when the previous conditions passes?

matsest replied:
Hm, no. The only requirement for the command to run is to have the terraform binary. It can run in all directories, even those who have nothing to do with Terraform. Even in directories that have not been subject to "terraform init" (i.e. they will not have a .terraform subdir) the command will output "default".

But it does not make sense to show this output in directories without the .terraform subdir, so that's why that check is added.

JanDeDobbeleer commented:
@matsest feel free to reach out for help when you need it, I'm here to assist. Great work overall, looking forward to merging this 👍🏻

JanDeDobbeleer commented on docs/docs/segment-terraform.md:
```diff
@@ -0,0 +1,28 @@
+  "foreground": "#000000",
+  "background": "#ebcc34",
+  "properties": {
+    "prefix": " "
```
This is the default so that can be omitted :-)

matsest replied:
Which part? Remove the properties :  {} statement? @JanDeDobbeleer

JanDeDobbeleer replied:
No, prefix defaults to `" "`

JanDeDobbeleer commented:
2 comments and a general remark, can you squash all these commits into 1? It's small enough to be in a single commit context.

matsest commented:
@JanDeDobbeleer Sounds good! Will fix this when I'm back home tomorrow evening

matsest commented:
Thanks for the helpful review. I've gone through your review comments, rebased and squashed the PR into one commit. Hope it looks good now @JanDeDobbeleer  :)

JanDeDobbeleer commented on docs/docs/segment-terraform.md:
```diff
@@ -0,0 +1,27 @@
+  "powerline_symbol": "\uE0B0",
+  "foreground": "#000000",
+  "background": "#ebcc34",
+  "properties": {
```
in this case this can be removed entirely :-)

JanDeDobbeleer commented:
> Hope it looks good now

Only need to remove the properties in the docs and it can be merged!
